### ISSUE TITLE
Optimizing an LLM Using DPO:  nan Loss Values During Evaluation

### ISSUE THREAD

**Issue by Refinath** [huggingface/trl#1789]
I want to optimize an LLM based on DPO. When I tried to train and evaluate the model, but there are nan values in the evaluation results.

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from datasets import Dataset
from trl import DPOTrainer, DPOConfig
from datasets import load_dataset

model_name = "EleutherAI/pythia-14m"
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

def preprocess_data(item):
    return {
        'prompt': 'Instruct: ' + item['prompt'] + '\n',
        'chosen': 'Output: ' + item['chosen'],
        'rejected': 'Output: ' + item['rejected']
    }

dataset = load_dataset('jondurbin/truthy-dpo-v0.1', split="train")
dataset = dataset.map(preprocess_data)
split_dataset = dataset.train_test_split(test_size=0.1)  # Adjust the test_size as needed
train_dataset = split_dataset['train']
val_dataset = split_dataset['test']

print(f"Length of train data: {len(train_dataset)}")
print(f"Length of validation data: {len(val_dataset)}")

# Tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name)
tokenizer.pad_token = tokenizer.unk_token

# Model to fine-tune
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    torch_dtype=torch.float16
).to(device)

model_ref = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    torch_dtype=torch.float16
).to(device)

# Config
training_args = DPOConfig(
    output_dir="./output",
    beta=0.1,
    max_length=512,
    max_prompt_length=128,
    remove_unused_columns=False,
)

# Load trainer
dpo_trainer = DPOTrainer(
    model,
    model_ref,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,  
    tokenizer=tokenizer,
)

# Train
dpo_trainer.train()

# Evaluate
evaluation_results = dpo_trainer.evaluate()
print("Evaluation Results:", evaluation_results)

This is the code used to train a simple 'pythia-14m' model. Below is the result.

Evaluation Results: {'eval_loss': nan, 'eval_runtime': 0.5616, 'eval_samples_per_second': 181.61, 'eval_steps_per_second': 12.463, 'eval_rewards/chosen': nan, 'eval_rewards/rejected': nan, 'eval_rewards/accuracies': 0.0, 'eval_rewards/margins': nan, 'eval_logps/rejected': nan, 'eval_logps/chosen': nan, 'eval_logits/rejected': nan, 'eval_logits/chosen': nan, 'epoch': 3.0}

any idea why nan values during evaluation ? is there anything wrong in the code ?

**Comment by qgallouedec** [comment]
Thanks for sharing this feedback.
Try with `torch_dtype=torch.bfloat16` instead.
Also, next time please use the markdown code block for the code
